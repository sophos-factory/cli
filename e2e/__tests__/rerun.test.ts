/* eslint-disable @typescript-eslint/no-non-null-assertion */

import isArray from 'lodash/isArray';
import { execute } from '../helpers/execute';
import knownIds from '../helpers/knownIds';

describe('refactrctl rerun', () => {
  // Give platform time to bootstrap runner.
  jest.setTimeout(60 * 1000);

  let runId: string;
  beforeAll(async () => {
    const data = JSON.parse(
      await execute(
        [
          'run',
          'pipeline',
          '--project-id',
          knownIds.dynamic.project,
          '--revision-id',
          knownIds.dynamic.pipelineRevision,
          knownIds.dynamic.pipeline,
          '--format=json'
        ],
        { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
      )
    );

    runId = data._id;

    // Wait until run is finished.
    return await execute(
      ['get', 'run', '--project-id', knownIds.dynamic.project, '--wait', runId],
      { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
    );
  });

  test('throws on missing arguments', async () => {
    await expect(
      execute(['run', 'pipeline'], {
        token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN!
      })
    ).rejects.toMatchSnapshot();
  });

  test('returns run object with rerunning without --wait flag', async () => {
    await expect(
      execute(
        [
          'rerun',
          '--project-id',
          knownIds.dynamic.project,
          runId,
          '--format=json'
        ],
        { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
      ).then((value) => JSON.parse(value))
    ).resolves.toHaveProperty('status', 'Queued');
  });

  test('returns array of events when running with --wait flag', async () => {
    await expect(
      execute(
        [
          'rerun',
          '--project-id',
          knownIds.dynamic.project,
          runId,
          '--wait',
          '--format=json'
        ],
        { token: process.env.FACTORY_DYNAMIC_AUTH_TOKEN! }
      ).then((value) => isArray(JSON.parse(value)))
    ).resolves.toBeTruthy();
  });
});
