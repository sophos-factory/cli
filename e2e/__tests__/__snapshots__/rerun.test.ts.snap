// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refactrctl rerun throws on missing arguments 1`] = `
"refactrctl run pipeline [pipeline-id]

Executes specified pipeline

Positionals:
  pipeline-id  ID of the pipeline to be executed  [string] [default: read from stdin]

Options:
      --version             Show version number  [boolean]
  -v, --verbose             Print detailed output  [boolean]
      --format              Output format  [choices: \\"wide\\", \\"json\\", \\"yaml\\"] [default: \\"wide\\"]
      --filter              Filter output using JsonPath  [string]
      --address             Address of the Sophos Factory API server  [string] [default: FACTORY_ADDRESS environment variable if set, otherwise https://api.refactr.it/v1]
      --auth-token          Authentication token  [string] [default: FACTORY_AUTH_TOKEN environment variable]
      --help                Show help  [boolean]
      --wait                Wait until run execution is finished  [boolean]
      --suppress-events     Suppress run events during pipeline or job run  [boolean]
      --suppress-outputs    Suppress outputs during pipeline or job run  [boolean]
      --suppress-variables  Suppress variables during pipeline or job run  [boolean]
      --var                 Pipeline variable in \`key:value\` format, where value is JSON data or valid JSON object.  [string]
      --project-id          ID of the project containing the pipeline  [string] [required]
      --revision-id         ID of the pipeline revision to be executed  [string] [required]

Missing required arguments: project-id, revision-id
"
`;
